collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -152,7 +151,140 @@ object ImmutableArray extends StrictOptimizedSeqFactory[ImmutableArray] {
       ScalaRunTime.array_update(elements, i, f(i))
       i = i + 1
     }
-    new ImmutableArray(elements)
+    ImmutableArray.wrapArray(elements)
+  }
+
+  def wrapArray[T](x: AnyRef): ImmutableArray[T] = (x match {
```

Comment:
Does this method have to be public? We should at least document what it does and also warn about the fact that it doesn’t create a copy of the given `Array`.

Reply:
It needs to be public in the bytecode so that the codegen can generate code that calls it, in cases that are known to be safe (i.e., varargs generation). However it could be private-ish at the language level. 

One way is to make it `protected[immutable]`.

Reply:
I don't think it will buy us anything to hide the method. The specialized subclasses expose the constructors and the wrapped arrays, and both are necessary to write high-performance code that uses ImmutableArray. Also exposing wrapArray on top of that does not make it any less safe. The current collections do the same and I have not seen any problems in the wild that were caused by this.

Reply:
Having it non-public would just prevent confusion.

Reply:
> The current collections do the same and I have not seen any problems in the wild that were caused by this.

Hum ... the current collections don't have `ImmutableArray`s to begin with, so I'm not sure what you're talking about.

Reply:
It's called `WrappedArray` in the current collections. (`WrappedArray` is a *mutable* collection though, so maybe it would be more problematic for `ImmutableArray`)

Reply:
> WrappedArray is a mutable collection though

Ah, well, yes. That changes everything!